python/cusignal/io/reader.py

```diff
@@ -0,0 +1,328 @@
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cupy as cp
+import numpy as np
+
+import json
+import resource
+
+from mmap import mmap, MAP_PRIVATE, PROT_READ
+
+from ._reader_cuda import _unpack, _pack
+
+
+# https://github.com/cupy/cupy/blob/master/examples/stream/cupy_memcpy.py
+def _pin_memory(array):
+    mem = cp.cuda.alloc_pinned_memory(array.nbytes)
+    ret = np.frombuffer(mem, array.dtype, array.size).reshape(array.shape)
+    ret[...] = array
+    return ret
+
+
+def pin_memory(size, dtype):
+    """
+    Create a pinned memory buffer.
+
+    Parameters
+    ----------
+    size : int or tuple of ints
+        Output shape.
+    dtype : data-type
+        Output data type.
+
+    Returns
+    -------
+    out : ndarray
+        Pinned memory numpy array.
+
+    """
+    pinned_memory_pool = cp.cuda.PinnedMemoryPool()
+    cp.cuda.set_pinned_memory_allocator(pinned_memory_pool.malloc)
+
+    x_cpu_dst = np.empty(size, dtype)
+    x_pinned_cpu_dst = _pin_memory(x_cpu_dst)
+
+    return x_pinned_cpu_dst
+
+
+def read_bin(file, dtype=None, num_samples=0, offset=0):
+    """
+    Reads binary file input GPU memory.
```

Comment:
Should this be 'reads binary final into GPU memory?'

Let's also add that this can be be used as a building block for reading other signal/file types.